### PR TITLE
[BEAM-222] Respect checkpointing contract in TestCountingSource

### PR DESCRIPTION
**Pull request by mshields822**
R: @dhalperi

### CODE REVIEW

**Comment by dhalperi** [comment]
test?

**Comment by mshields822** [comment]
I'm going to follow your advice in the issue and nuke this entirely. Stand by.

**Comment by mshields822** [comment]
It's Friday, I'm already 3 steps removed from the problem I set out to solve.
Adding a test and leaving this in place.

**Comment by dhalperi** [comment]
Looks like the test is flaky:

```
Tests run: 4, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 12.976 sec <<< FAILURE! - in org.apache.beam.sdk.io.BoundedReadFromUnboundedSourceTest
testDedup(org.apache.beam.sdk.io.BoundedReadFromUnboundedSourceTest)  Time elapsed: 0.034 sec  <<< FAILURE!
java.lang.AssertionError: expected:<0> but was:<-1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:743)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at org.junit.Assert.assertEquals(Assert.java:555)
	at org.junit.Assert.assertEquals(Assert.java:542)
	at org.apache.beam.sdk.io.BoundedReadFromUnboundedSourceTest$Checker.apply(BoundedReadFromUnboundedSourceTest.java:113)
	at org.apache.beam.sdk.io.BoundedReadFromUnboundedSourceTest$Checker.apply(BoundedReadFromUnboundedSourceTest.java:85)
	at org.apache.beam.sdk.testing.PAssert$CheckerDoFn.processElement(PAssert.java:627)
	at org.apache.beam.sdk.util.SimpleDoFnRunner.invokeProcessElement(SimpleDoFnRunner.java:51)
	at org.apache.beam.sdk.util.DoFnRunnerBase.processElement(DoFnRunnerBase.java:140)
	at org.apache.beam.sdk.transforms.ParDo.evaluateHelper(ParDo.java:1229)
	at org.apache.beam.sdk.transforms.ParDo.evaluateSingleHelper(ParDo.java:1098)
	at org.apache.beam.sdk.transforms.ParDo.access$300(ParDo.java:459)
	at org.apache.beam.sdk.transforms.ParDo$1.evaluate(ParDo.java:1084)
	at org.apache.beam.sdk.transforms.ParDo$1.evaluate(ParDo.java:1079)
	at org.apache.beam.sdk.runners.DirectPipelineRunner$Evaluator.visitTransform(DirectPipelineRunner.java:904)
	at org.apache.beam.sdk.runners.TransformTreeNode.visit(TransformTreeNode.java:221)
	at org.apache.beam.sdk.runners.TransformTreeNode.visit(TransformTreeNode.java:217)
	at org.apache.beam.sdk.runners.TransformTreeNode.visit(TransformTreeNode.java:217)
	at org.apache.beam.sdk.runners.TransformHierarchy.visit(TransformHierarchy.java:104)
	at org.apache.beam.sdk.Pipeline.traverseTopologically(Pipeline.java:261)
	at org.apache.beam.sdk.runners.DirectPipelineRunner$Evaluator.run(DirectPipelineRunner.java:860)
	at org.apache.beam.sdk.runners.DirectPipelineRunner.run(DirectPipelineRunner.java:572)
	at org.apache.beam.sdk.runners.DirectPipelineRunner.run(DirectPipelineRunner.java:106)
	at org.apache.beam.sdk.Pipeline.run(Pipeline.java:182)
	at org.apache.beam.sdk.testing.TestPipeline.run(TestPipeline.java:111)
	at org.apache.beam.sdk.io.BoundedReadFromUnboundedSourceTest.test(BoundedReadFromUnboundedSourceTest.java:147)
	at org.apache.beam.sdk.io.BoundedReadFromUnboundedSourceTest.testDedup(BoundedReadFromUnboundedSourceTest.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runners.Suite.runChild(Suite.java:127)
	at org.junit.runners.Suite.runChild(Suite.java:26)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:138)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:108)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:78)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:54)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:144)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray2(ReflectionUtils.java:208)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:155)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:81)
	at org.apache.maven.plugin.surefire.InPluginVMSurefireStarter.runSuitesInProcess(InPluginVMSurefireStarter.java:82)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:974)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:824)
	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:722)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
```

from [this log](https://s3.amazonaws.com/archive.travis-ci.org/jobs/125160420/log.txt)

**Comment by dhalperi** [comment]
Sorry, more accurately, looks like this may have caused a different test using this source to become flaky.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java [hunk 1]
```diff
@@ -164,21 +164,20 @@ public CountingSourceReader(int startingPoint) {
...
         // If testing dedup, occasionally insert a duplicate value;
         if (dedup && ThreadLocalRandom.current().nextInt(5) == 0) {
           return true;
         }
         current++;
-        return true;
-      } else {
-        return false;
-      }
+      return current < numMessagesPerShard;
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: the indentation looks off here, but maybe it's a tabs-vs-spaces thing? Except we shouldn't have tabs.. hmm.. can you please investigate?

Comment (mshields822) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: Fixed.

**Comment by mshields822** [comment]
Fixed flaky test - it was a real bug that I introduced so that's good.

**Comment by mshields822** [comment]
R: @dhalperi 
PTAL

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSourceTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Test the TestCountingSource.
+ */
+@RunWith(JUnit4.class)
+public class TestCountingSourceTest {
+  @Test
+  public void testRespectsCheckpointContract() throws IOException {
+    TestCountingSource source = new TestCountingSource(3);
+    PipelineOptions options = PipelineOptionsFactory.create();
+    TestCountingSource.CountingSourceReader reader = source.createReader(options, null);
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSourceTest.java]: null -> null /* no checkpoint */

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java [hunk 3]
```diff
@@ -164,21 +168,20 @@ public CountingSourceReader(int startingPoint) {
...
-          return true;
-        }
-        current++;
-        return true;
-      } else {
+      if (current >= numMessagesPerShard) {
         return false;
       }
+      // If testing dedup, occasionally insert a duplicate value;
+      if (current >= 0 && dedup && ThreadLocalRandom.current().nextInt(5) == 0) {
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: Won't this produce duplicates before the checkpoint mark?

So we will output 0, then 1, then checkpoint that we just output 1, and then we will potentially output 1 again after restoring from checkpoint?

It seems as though restoring from checkpoint should guarantee that the previously emit values won't be emit again.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java [hunk 4]
```diff
@@ -164,21 +168,20 @@ public CountingSourceReader(int startingPoint) {
 
     @Override
     public boolean start() {
-      return true;
+      return advance();
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: Contractually you can only be called to start once, use another variable to track whether you have been started and throw an error if you have been called multiple times.

Also, check that in advance, that start has been called.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java [hunk 5]
```diff
@@ -164,21 +168,20 @@ public CountingSourceReader(int startingPoint) {
...
       }
+      // If testing dedup, occasionally insert a duplicate value;
+      if (current >= 0 && dedup && ThreadLocalRandom.current().nextInt(5) == 0) {
+        return true;
+      }
+      current++;
+      return current < numMessagesPerShard;
     }
 
     @Override
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: Ensure that all the get/close calls ensure that start has been called.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java (hunk 3 above)
Comment (mshields822) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: Yep, the implementation is broken, but already was broken.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java (hunk 4 above)
Comment (mshields822) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: This guy is on the way out - it should be replaced by CountingSource, so I'm not going to go to town here.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java (hunk 5 above)
Comment (mshields822) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/runners/dataflow/TestCountingSource.java]: As above.

**Comment by lukecwik** [comment]
LGTM

Need to fund getting rid of TestCountingSource in favor of CountingSource and having a test artifact

**Comment by mshields822** [comment]
Agree.

**Pull request merged**